managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -1850,9 +1854,172 @@ public void offloadFailed(ManagedLedgerException e, Object ctx) {
+                    List<Pair<Long, byte[]>> successfulOffloads = Lists.newArrayList();
+                    int errors = 0;
+                    synchronized (this) {
+                        ledgersListMutex.lock();
+
+                        // loop through results of offload operations. If an error occurred
+                        // check that the ledger still exists for the ML. If not, then it was
+                        // trimmed and the error can be ignored.
+                        for (Pair<Long, CompletableFuture<byte[]>> context : contexts) {
+                            if (context.getRight().isCompletedExceptionally()) {
+                                if (ledgers.containsKey(context.getLeft())) {
+                                    errors++;
+                                }
+                            } else {
+                                successfulOffloads.add(Pair.of(context.getLeft(), context.getRight().join()));
+                            }
+                        }
+                        ledgersListMutex.unlock();
+                    }
+                    log.info("[{}] All offload operations complete, {} successful, {} errors",
+                             name, successfulOffloads.size(), errors);
+
+                    int errorsToReport = errors; // effectively final so can be used in lambda
+                    if (successfulOffloads.size() > 0) {
+                        updateLedgerInfoForOffloaded(successfulOffloads).whenComplete(
+                                (ignore2, exception) -> {
+                                    if (exception != null) {
+                                        callback.offloadFailed(new ManagedLedgerException(exception), ctx);
+                                    } else if (offloadException != null && errorsToReport > 0) {
+                                        callback.offloadFailed(new ManagedLedgerException(offloadException), ctx);
+                                    } else {
+                                        callback.offloadComplete(firstUnoffloaded, ctx);
+                                    }
+                                });
+                    } else {
+                        // All failed, so throw up the error
+                        callback.offloadFailed(new ManagedLedgerException(offloadException), ctx);
+                    }
+                });
     }
 
+    private CompletableFuture<Void> updateLedgerInfoForOffloaded(List<Pair<Long, byte[]>> contexts) {
+        CompletableFuture<Void> promise = new CompletableFuture<>();
+        synchronized (this) {
+            ledgersListMutex.lock();
+
+            contexts.forEach((context) -> {
+                    long ledgerId = context.getLeft();
+                    LedgerInfo oldInfo = ledgers.get(ledgerId);
+                    if (oldInfo == null) {
+                        cleanupOffloadedOnFailure(context.getLeft(), context.getRight(),
```

Comment:
if pulsar broker fails during offloading, who is responsible for cleaning up those "partially completed" ledgers in tier storage?

I was imaging a sequence like below:

- update metadata indicate a ledger is offloading
- offloading
- update metadata indicate a ledger is offloaded

so if a ML is crashed during offloading, when it is recovered and start offloading again, it will know the state from ledger info. however I don't see metadata being updated before offloading, do I miss anything here?

Reply:
My thinking on it was that there'd be a GC type mechanism as crashes during offload should be pretty rare, similar to what must happen if a ledger is created but the ML crashes before updating the managed ledger metadata.

Being able to cleanup a failed offload implies the location being deterministic from the ledger info. We should avoid requiring that the name is deterministic from the ledger ids, as then we could end up with nasty races, where one is broker is offloading and pauses, and another picks up the offloading process, then the first one wakes up and corrupts the offloading process. Unlikely, but there's no fencing, so entirely possible.

We could do it by adding an offloadId UUID parameter to LedgerOffloader calls. We could generate this, store in offload indication field, offload, and the update context to indictate that offload is complete. 

This change would also imply more writes on ZK, which may not be desirable. @merlimat What are your thoughts on this?

Reply:
> as crashes during offload should be pretty rare, similar to what must happen if a ledger is created but the ML crashes before updating the managed ledger metadata.

these two cases are different in terms of costs. 

" ledger is created but the ML crashes before updating the managed ledger metadata" => there is no data in the ledger. so it is effectively just one metadata.

"crashes during offload" => it can happen after you fully copy the data. so you are leaving a bucket in tiered storage.

> Being able to cleanup a failed offload implies the location being deterministic from the ledger info.

It doesn't need to a deterministic name. it just need to be tracked somewhere for cleanup. `somewhere` can be the ledger info in ML.

Reply:
I'll change the offloader to take a UUID, and remove the context stuff (since it's not useful since we wouldn't be able to use it to find the object since we need to know all parameters before creating it).